src/main/java/io/github/mirromutth/r2dbc/mysql/WindowTinyLfuQueryCache.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2018-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.github.mirromutth.r2dbc.mysql;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
+
+import static io.github.mirromutth.r2dbc.mysql.internal.AssertUtils.require;
+
+/**
+ * An implementation of {@link QueryCache} which based on Window Tiny LFU.
+ * <p>
+ * The page replacement algorithm has kept eventual consistency with the hashing map like Caffeine,
+ * but using a different and cleanly implementation based on special features, that are unable to remove
+ * any element by user actively, and only load the element when it does not exist in the cache.
+ * <p>
+ * In other words, this W-TinyLFU implementation has only {@literal computeIfAbsent} and highly optimized
+ * by this feature. So it will never support actively getIfPresent/putOverlay/removeIfPresent until
+ * refactor to other implementation.
+ * <p>
+ * It is thread-safety.
+ *
+ * @see #consume(Lru.Node) only remove element in this method.
+ */
+final class WindowTinyLfuQueryCache implements QueryCache {
+
+    private static final int CAS_TIMES = 100;
+
+    private static final Function<String, Lru.Node<Query>> COMPUTER = sql -> new Lru.Node<>(sql, Query.parse(sql));
+
+    private final ConcurrentMap<String, Lru.Node<Query>> hashing = new ConcurrentHashMap<>();
+
+    private final Lock lock = new ReentrantLock();
+
+    private final Lru<Query> window;
+
+    private final Slru<Query> slru;
+
+    private final FrequencySketch sketch;
+
+    private final RingBuffer<Lru.Node<Query>> buffer;
+
+    WindowTinyLfuQueryCache(int capacity) {
+        require(capacity > 0, "capacity must be a positive integer");
+
+        int windowSize = capacity / 100;
+
+        this.window = new Lru<>(Math.max(1, windowSize));
+        this.slru = new Slru<>(Math.max(1, capacity - windowSize));
+        this.sketch = new FrequencySketch(capacity);
+        this.buffer = new RingBuffer<>(this::consume);
+    }
+
+    @Override
+    public Query getOrParse(String sql) {
+        Lru.Node<Query> node = hashing.computeIfAbsent(sql, COMPUTER);
+        push(node);
+        return node.getValue();
+    }
+
+    private void push(Lru.Node<Query> node) {
```

Comment:
You may want to break into apart into separate read/write buffers, like Caffeine does.
 - When a cache miss occurs then the entry must be added to the LRU and maybe an eviction occurs. This must occur, so it has the work has to be processed and not dropped.
 - When a cache hit occurs then processing the node is a policy hint, e.g. to LRU reorder. This is best-effort and the work could be dropped, e.g. if the buffer is full. This favors concurrent throughput and likely the same popular entries are being accessed. If so, the eviction policy is already going to favor them so being strict won't change the hit rate substantially. Being lossy can improve the system throughput, so generally a favorable tradeoff.